// @crystallize/js-api-client/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CrystallizeCustomerManager = exports.CrystallizeSubscriptionContractManager = exports.CrystallizeOrderFetcherByCustomerIdentifier = exports.CrystallizeOrderFetcherById = exports.CrystallizeSearcher = exports.CrystallizeCatalogueFetcher = exports.CrystallizeCreateOrderPaymentUpdater = exports.CrystallizeOrderPusher = exports.CrystallizeHydraterBySkus = exports.CrystallizeHydraterByPaths = exports.CrystallizeNavigationTopicsFetcher = exports.CrystallizeNavigationFoldersFetcher = exports.CrystallizeClient = void 0;
__exportStar(require("./core/client"), exports);
__exportStar(require("./core/massCallClient"), exports);
__exportStar(require("./core/navigation"), exports);
__exportStar(require("./core/hydrate"), exports);
__exportStar(require("./core/catalogue"), exports);
__exportStar(require("./core/order"), exports);
__exportStar(require("./core/search"), exports);
__exportStar(require("./core/subscription"), exports);
__exportStar(require("./core/customer"), exports);
__exportStar(require("./core/pricing"), exports);
__exportStar(require("./core/verifySignature"), exports);
__exportStar(require("./types/product"), exports);
__exportStar(require("./types/order"), exports);
__exportStar(require("./types/payment"), exports);
__exportStar(require("./types/components"), exports);
__exportStar(require("./types/search"), exports);
__exportStar(require("./types/subscription"), exports);
__exportStar(require("./types/address"), exports);
__exportStar(require("./types/customer"), exports);
__exportStar(require("./types/signature"), exports);
__exportStar(require("./types/pricing"), exports);
const client_1 = require("./core/client");
const navigation_1 = require("./core/navigation");
const hydrate_1 = require("./core/hydrate");
const order_1 = require("./core/order");
const catalogue_1 = require("./core/catalogue");
const search_1 = require("./core/search");
const subscription_1 = require("./core/subscription");
const customer_1 = require("./core/customer");
exports.CrystallizeClient = (0, client_1.createClient)({
    tenantId: globalThis?.process?.env?.CRYSTALLIZE_TENANT_ID ?? '',
    tenantIdentifier: globalThis?.process?.env?.CRYSTALLIZE_TENANT_IDENTIFIER ?? '',
    accessTokenId: globalThis?.process?.env?.CRYSTALLIZE_ACCESS_TOKEN_ID ?? '',
    accessTokenSecret: globalThis?.process?.env?.CRYSTALLIZE_ACCESS_TOKEN_SECRET ?? '',
});
const navigationFetcher = (0, navigation_1.createNavigationFetcher)(exports.CrystallizeClient);
exports.CrystallizeNavigationFoldersFetcher = navigationFetcher.byFolders;
exports.CrystallizeNavigationTopicsFetcher = navigationFetcher.byTopics;
const productHydrator = (0, hydrate_1.createProductHydrater)(exports.CrystallizeClient);
exports.CrystallizeHydraterByPaths = productHydrator.byPaths;
exports.CrystallizeHydraterBySkus = productHydrator.bySkus;
exports.CrystallizeOrderPusher = (0, order_1.createOrderPusher)(exports.CrystallizeClient);
exports.CrystallizeCreateOrderPaymentUpdater = (0, order_1.createOrderPaymentUpdater)(exports.CrystallizeClient);
exports.CrystallizeCatalogueFetcher = (0, catalogue_1.createCatalogueFetcher)(exports.CrystallizeClient);
exports.CrystallizeSearcher = (0, search_1.createSearcher)(exports.CrystallizeClient);
const orderFetcher = (0, order_1.createOrderFetcher)(exports.CrystallizeClient);
exports.CrystallizeOrderFetcherById = orderFetcher.byId;
exports.CrystallizeOrderFetcherByCustomerIdentifier = orderFetcher.byCustomerIdentifier;
exports.CrystallizeSubscriptionContractManager = (0, subscription_1.createSubscriptionContractManager)(exports.CrystallizeClient);
exports.CrystallizeCustomerManager = (0, customer_1.createCustomerManager)(exports.CrystallizeClient);
