// @crystallize/js-api-client/dist/types/address.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addressInputRequest = void 0;
const zod_1 = require("zod");
const json_to_graphql_query_1 = require("json-to-graphql-query");
exports.addressInputRequest = zod_1.z
    .object({
    type: zod_1.z.enum(['delivery', 'billing', 'other']).transform((val) => new json_to_graphql_query_1.EnumType(val)),
    firstName: zod_1.z.string().optional(),
    middleName: zod_1.z.string().optional(),
    lastName: zod_1.z.string().optional(),
    street: zod_1.z.string().optional(),
    street2: zod_1.z.string().optional(),
    streetNumber: zod_1.z.string().optional(),
    postalCode: zod_1.z.string().optional(),
    city: zod_1.z.string().optional(),
    state: zod_1.z.string().optional(),
    country: zod_1.z.string().optional(),
    phone: zod_1.z.string().optional(),
    email: zod_1.z.string().optional(),
    meta: zod_1.z.array(zod_1.z.object({ key: zod_1.z.string(), value: zod_1.z.string().optional() })).optional(),
})
    .strict();
