// @crystallize/js-api-client/dist/types/customer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.updateCustomerInputRequest = exports.createCustomerInputRequest = exports.orderCustomerInputRequest = void 0;
const zod_1 = require("zod");
const address_1 = require("./address");
exports.orderCustomerInputRequest = zod_1.z
    .object({
    identifier: zod_1.z.string().optional(),
    firstName: zod_1.z.string().optional(),
    middleName: zod_1.z.string().optional(),
    lastName: zod_1.z.string().optional(),
    birthDate: zod_1.z.date().optional(),
    companyName: zod_1.z.string().optional(),
    taxNumber: zod_1.z.string().optional(),
    addresses: zod_1.z.array(address_1.addressInputRequest).optional(),
})
    .strict();
exports.createCustomerInputRequest = exports.orderCustomerInputRequest
    .extend({
    tenantId: zod_1.z.string().optional(),
    lastName: zod_1.z.string(),
    firstName: zod_1.z.string(),
    phone: zod_1.z.string().optional(),
    meta: zod_1.z
        .array(zod_1.z.object({
        key: zod_1.z.string(),
        value: zod_1.z.string().optional(),
    }))
        .optional(),
    identifier: zod_1.z.string().optional(),
    externalReferences: zod_1.z
        .array(zod_1.z.object({
        key: zod_1.z.string(),
        value: zod_1.z.string().optional(),
    }))
        .optional(),
    email: zod_1.z.string(),
})
    .strict();
exports.updateCustomerInputRequest = exports.createCustomerInputRequest.omit({ identifier: true, tenantId: true });
