// @crystallize/js-api-client/dist/types/payment.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.customPaymentInputRequest = exports.cashPaymentInputRequest = exports.stripePaymentInputRequest = exports.paypalPaymentInputRequest = exports.klarnaPaymentInputRequest = exports.paymentProvider = void 0;
const zod_1 = require("zod");
const json_to_graphql_query_1 = require("json-to-graphql-query");
exports.paymentProvider = zod_1.z
    .enum(['klarna', 'stripe', 'paypal', 'cash', 'custom'])
    .transform((val) => new json_to_graphql_query_1.EnumType(val));
// Klarna
exports.klarnaPaymentInputRequest = zod_1.z
    .object({
    klarna: zod_1.z.string().optional(),
    orderId: zod_1.z.string().optional(),
    recurringToken: zod_1.z.string().optional(),
    status: zod_1.z.string().optional(),
    merchantReference1: zod_1.z.string().optional(),
    merchantReference2: zod_1.z.string().optional(),
    metadata: zod_1.z.string().optional(),
})
    .strict();
// Paypal
exports.paypalPaymentInputRequest = zod_1.z
    .object({
    paypal: zod_1.z.string().optional(),
    orderId: zod_1.z.string().optional(),
    subscriptionId: zod_1.z.string().optional(),
    invoiceId: zod_1.z.string().optional(),
    metadata: zod_1.z.string().optional(),
})
    .strict();
// Stripe
exports.stripePaymentInputRequest = zod_1.z
    .object({
    stripe: zod_1.z.string().optional(),
    customerId: zod_1.z.string().optional(),
    orderId: zod_1.z.string().optional(),
    paymentMethod: zod_1.z.string().optional(),
    paymentMethodId: zod_1.z.string().optional(),
    paymentIntentId: zod_1.z.string().optional(),
    subscriptionId: zod_1.z.string().optional(),
    metadata: zod_1.z.string().optional(),
})
    .strict();
// Cash
exports.cashPaymentInputRequest = zod_1.z
    .object({
    cash: zod_1.z.string().optional(),
})
    .strict();
// Custom
exports.customPaymentInputRequest = zod_1.z
    .object({
    properties: zod_1.z
        .array(zod_1.z.object({
        property: zod_1.z.string(),
        value: zod_1.z.string().optional(),
    }))
        .optional(),
})
    .strict();
