// @crystallize/js-api-client/dist/types/search.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.catalogueSearchOrderBy = exports.catalogueSearchFilter = void 0;
const json_to_graphql_query_1 = require("json-to-graphql-query");
const zod_1 = require("zod");
const topicPathsFilterField = zod_1.z
    .object({
    value: zod_1.z.string(),
})
    .strict();
const topicPathsFilterSection = zod_1.z
    .object({
    logicalOperator: zod_1.z.enum(['AND', 'OR']).transform((val) => new json_to_graphql_query_1.EnumType(val)),
    fields: zod_1.z.array(topicPathsFilterField).optional(),
})
    .strict();
const topicPathsFilter = zod_1.z
    .object({
    logicalOperator: zod_1.z.enum(['AND', 'OR']).transform((val) => new json_to_graphql_query_1.EnumType(val)),
    sections: zod_1.z.array(topicPathsFilterSection),
})
    .strict();
const priceRangeFilter = zod_1.z
    .object({
    min: zod_1.z.number(),
    max: zod_1.z.number(),
})
    .strict();
const stockFilter = zod_1.z
    .object({
    min: zod_1.z.number(),
    location: zod_1.z.string().optional(),
})
    .strict();
const stockLocationsFilter = zod_1.z
    .object({
    min: zod_1.z.number(),
    location: zod_1.z.array(zod_1.z.string()).optional(),
    logicalOperator: zod_1.z.enum(['OR']),
})
    .strict();
const itemFilterFields = zod_1.z
    .object({
    itemIds: zod_1.z.string().optional(),
    productVariantIds: zod_1.z.string().optional(),
    skus: zod_1.z.string().optional(),
    shapeIdentifiers: zod_1.z.string().optional(),
    paths: zod_1.z.string().optional(),
    topicsPaths: topicPathsFilter.optional(),
})
    .strict();
const variantAttributeFilter = zod_1.z
    .object({
    attribute: zod_1.z.string(),
    value: zod_1.z.string(),
})
    .strict();
const productVariantsFilter = zod_1.z.object({
    isDefault: zod_1.z.boolean().optional(),
    priceRange: priceRangeFilter.optional(),
    stock: stockFilter.optional(),
    stockLocations: stockLocationsFilter.optional(),
    attributes: variantAttributeFilter.optional(),
});
exports.catalogueSearchFilter = zod_1.z.object({
    searchTerm: zod_1.z.string().optional(),
    type: zod_1.z
        .enum(['PRODUCT', 'FOLDER', 'DOCUMENT'])
        .transform((val) => new json_to_graphql_query_1.EnumType(val))
        .optional(),
    include: itemFilterFields.optional(),
    exclude: itemFilterFields.optional(),
    priceVariant: zod_1.z.string().optional(),
    stockLocation: zod_1.z.string().optional(),
    productVariants: productVariantsFilter.optional(),
});
exports.catalogueSearchOrderBy = zod_1.z
    .object({
    field: zod_1.z.enum(['ITEM_NAME', 'PRICE', 'STOCK', 'CREATED_AT']).transform((val) => new json_to_graphql_query_1.EnumType(val)),
    direction: zod_1.z.enum(['ASC', 'DESC']).transform((val) => new json_to_graphql_query_1.EnumType(val)),
})
    .strict();
