// @crystallize/js-api-client/dist/core/catalogue.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.catalogueFetcherGraphqlBuilder = exports.createCatalogueFetcher = void 0;
const json_to_graphql_query_1 = require("json-to-graphql-query");
const catalogue_1 = require("../types/catalogue");
function createCatalogueFetcher(client) {
    return (query, variables) => {
        return client.catalogueApi((0, json_to_graphql_query_1.jsonToGraphQLQuery)({ query }), variables);
    };
}
exports.createCatalogueFetcher = createCatalogueFetcher;
exports.catalogueFetcherGraphqlBuilder = {
    onItem,
    onProduct,
    onDocument,
    onFolder,
    onComponent,
    onSubscriptionPlan,
};
function onItem(onItem, c) {
    return {
        __typeName: 'Item',
        __typename: true,
        name: true,
        path: true,
        ...onItem,
        topics: {
            name: true,
            path: true,
            ...(c?.onTopic ? c.onTopic : {}),
        },
    };
}
function onDocument(onDocument, c) {
    return {
        __typeName: 'Document',
        __typename: true,
        ...onDocument,
    };
}
function onFolder(onFolder, c) {
    const children = () => {
        if (c?.onChildren) {
            return {
                chidlren: {
                    ...c.onChildren,
                },
            };
        }
        return {};
    };
    return {
        __typeName: 'Folder',
        __typename: true,
        ...onFolder,
        ...children(),
    };
}
function onProduct(onProduct, c) {
    const priceVariant = () => {
        if (c?.onPriceVariant) {
            return {
                priceVariants: {
                    ...c.onPriceVariant,
                },
            };
        }
        return {};
    };
    const variants = () => {
        if (c?.onVariant) {
            return {
                variants: {
                    name: true,
                    sku: true,
                    price: true,
                    ...priceVariant(),
                    ...(c?.onVariant ? c.onVariant : {}),
                },
            };
        }
        return {};
    };
    const defaultVariant = () => {
        if (c?.onDefaultVariant) {
            return {
                defaultVariant: {
                    ...c.onDefaultVariant,
                },
            };
        }
        return {};
    };
    return {
        __typeName: 'Product',
        __typename: true,
        ...onProduct,
        vatType: {
            name: true,
            percent: true,
        },
        ...defaultVariant(),
        ...variants(),
    };
}
/**
 * Convert hyphenated string to camel cased string
 * @param id A string with potentially hyphens in Item
 * @returns A camel cased string
 */
const camelCaseHyphens = (id) => id.replace(/-([a-z])/g, (g) => g[1].toUpperCase());
function onComponent(id, type, onComponent, c) {
    const validType = catalogue_1.componentType.parse(type);
    const aliasName = camelCaseHyphens(id);
    return {
        [aliasName]: {
            __aliasFor: 'component',
            __args: {
                id,
            },
            content: {
                __typename: true,
                __on: {
                    __typeName: validType,
                    ...onComponent,
                },
            },
        },
    };
}
function onSubscriptionPlan(c) {
    const period = (name) => {
        return {
            ...(c?.onPeriod ? c.onPeriod(name) : {}),
            priceVariants: {
                identifier: true,
                name: true,
                price: true,
                currency: true,
            },
            meteredVariables: {
                id: true,
                name: true,
                identifier: true,
                tierType: true,
                tiers: {
                    threshold: true,
                    priceVariants: {
                        identifier: true,
                        name: true,
                        price: true,
                        currency: true,
                    },
                },
            },
        };
    };
    return {
        subscriptionPlans: {
            identifier: true,
            name: true,
            periods: {
                id: true,
                name: true,
                initial: period('initial'),
                recurring: period('recurring'),
            },
        },
    };
}
