// @crystallize/js-api-client/dist/core/client.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createClient = void 0;
const node_fetch_1 = __importDefault(require("node-fetch"));
function authenticationHeaders(config) {
    if (config.sessionId) {
        return {
            Cookie: 'connect.sid=' + config.sessionId,
        };
    }
    if (config.staticAuthToken) {
        return {
            'X-Crystallize-Static-Auth-Token': config.staticAuthToken,
        };
    }
    return {
        'X-Crystallize-Access-Token-Id': config.accessTokenId || '',
        'X-Crystallize-Access-Token-Secret': config.accessTokenSecret || '',
    };
}
async function post(path, config, query, variables, init, profiling) {
    try {
        const commonHeaders = {
            'Content-type': 'application/json; charset=UTF-8',
            Accept: 'application/json',
        };
        const headers = {
            ...commonHeaders,
            ...authenticationHeaders(config),
        };
        const body = JSON.stringify({ query, variables });
        let start = 0;
        if (profiling) {
            start = Date.now();
            if (profiling.onRequest) {
                profiling.onRequest(query, variables);
            }
        }
        const response = await (0, node_fetch_1.default)(path, {
            ...init,
            method: 'POST',
            headers,
            body,
        });
        if (profiling) {
            const ms = Date.now() - start;
            const serverTiming = response.headers.get('server-timing') ?? undefined;
            const duration = serverTiming?.split(';')[1]?.split('=')[1] ?? -1;
            profiling.onRequestResolved({
                resolutionTimeMs: ms,
                serverTimeMs: Number(duration),
            }, query, variables);
        }
        if (response.ok && 204 === response.status) {
            return {};
        }
        if (!response.ok) {
            const json = await response.json();
            throw {
                code: response.status,
                statusText: response.statusText,
                message: json.message,
                errors: json.errors || {},
            };
        }
        // we still need to check for error as the API can return 200 with errors
        const json = await response.json();
        if (json.errors) {
            throw {
                code: 400,
                statusText: 'Error was returned from the API',
                message: json.errors[0].message,
                errors: json.errors || {},
            };
        }
        return json.data;
    }
    catch (exception) {
        throw exception;
    }
}
function createApiCaller(uri, configuration, options) {
    return function callApi(query, variables) {
        return post(uri, configuration, query, variables, undefined, options?.profiling);
    };
}
function createClient(configuration, options) {
    const identifier = configuration.tenantIdentifier;
    const origin = configuration.origin || '.crystallize.com';
    const apiHost = (path, prefix = 'api') => `https://${prefix}${origin}/${path.join('/')}`;
    return {
        catalogueApi: createApiCaller(apiHost([identifier, 'catalogue']), configuration, options),
        searchApi: createApiCaller(apiHost([identifier, 'search']), configuration, options),
        orderApi: createApiCaller(apiHost([identifier, 'orders']), configuration, options),
        subscriptionApi: createApiCaller(apiHost([identifier, 'subscriptions']), configuration, options),
        pimApi: createApiCaller(apiHost(['graphql'], 'pim'), configuration, options),
        config: {
            tenantId: configuration.tenantId,
            tenantIdentifier: configuration.tenantIdentifier,
            origin: configuration.origin,
        },
    };
}
exports.createClient = createClient;
