// @crystallize/js-api-client/dist/core/order.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createOrderPaymentUpdater = exports.createOrderPusher = exports.createOrderFetcher = void 0;
const order_1 = require("../types/order");
const json_to_graphql_query_1 = require("json-to-graphql-query");
function buildQuery(onCustomer, onOrderItem, extraQuery) {
    return {
        id: true,
        createdAt: true,
        updatedAt: true,
        customer: {
            identifier: true,
            ...(onCustomer !== undefined ? onCustomer : {}),
        },
        cart: {
            name: true,
            sku: true,
            imageUrl: true,
            quantity: true,
            ...(onOrderItem !== undefined ? onOrderItem : {}),
            price: {
                gross: true,
                net: true,
                discounts: {
                    percent: true,
                },
            },
        },
        total: {
            gross: true,
            net: true,
            currency: true,
            discounts: {
                percent: true,
            },
            tax: {
                name: true,
                percent: true,
            },
        },
        ...(extraQuery !== undefined ? extraQuery : {}),
    };
}
function createOrderFetcher(apiClient) {
    // we don't provide the current cursor for each Order.
    const fetchPaginatedOrdersByCustomerIdentifier = async (customerIdentifier, extraQueryArgs, onCustomer, onOrderItem, extraQuery) => {
        const orderApi = apiClient.orderApi;
        const query = {
            orders: {
                getAll: {
                    __args: {
                        customerIdentifier: customerIdentifier,
                        ...(extraQueryArgs !== undefined ? extraQueryArgs : {}),
                    },
                    pageInfo: {
                        hasPreviousPage: true,
                        hasNextPage: true,
                        startCursor: true,
                        endCursor: true,
                        totalNodes: true,
                    },
                    edges: {
                        cursor: true,
                        node: buildQuery(onCustomer, onOrderItem, extraQuery),
                    },
                },
            },
        };
        const response = await orderApi((0, json_to_graphql_query_1.jsonToGraphQLQuery)({ query }));
        return {
            pageInfo: response.orders.getAll.pageInfo,
            orders: response.orders.getAll?.edges?.map((edge) => edge.node) || [],
        };
    };
    const fetchOrderById = async (orderId, onCustomer, onOrderItem, extraQuery) => {
        const orderApi = apiClient.orderApi;
        const query = {
            orders: {
                get: {
                    __args: {
                        id: orderId,
                    },
                    id: true,
                    createdAt: true,
                    updatedAt: true,
                    customer: {
                        identifier: true,
                        ...(onCustomer !== undefined ? onCustomer : {}),
                    },
                    cart: {
                        name: true,
                        sku: true,
                        imageUrl: true,
                        quantity: true,
                        ...(onOrderItem !== undefined ? onOrderItem : {}),
                        price: {
                            gross: true,
                            net: true,
                            discounts: {
                                percent: true,
                            },
                        },
                    },
                    total: {
                        gross: true,
                        net: true,
                        currency: true,
                        discounts: {
                            percent: true,
                        },
                        tax: {
                            name: true,
                            percent: true,
                        },
                    },
                    ...(extraQuery !== undefined ? extraQuery : {}),
                },
            },
        };
        return (await orderApi((0, json_to_graphql_query_1.jsonToGraphQLQuery)({ query })))?.orders?.get;
    };
    return {
        byId: fetchOrderById,
        byCustomerIdentifier: fetchPaginatedOrdersByCustomerIdentifier,
    };
}
exports.createOrderFetcher = createOrderFetcher;
function convertDates(intent) {
    if (!intent.cart) {
        return {
            ...intent,
        };
    }
    return {
        ...intent,
        cart: intent.cart.map((item) => {
            if (!item.subscription) {
                return {
                    ...item,
                };
            }
            return {
                ...item,
                subscription: {
                    ...item.subscription,
                    start: item.subscription.start?.toISOString(),
                    end: item.subscription.end?.toISOString(),
                },
            };
        }),
    };
}
function createOrderPusher(apiClient) {
    return async function pushOrder(intentOrder) {
        const intent = order_1.createOrderInputRequest.parse(intentOrder);
        const orderApi = apiClient.orderApi;
        const mutation = {
            mutation: {
                orders: {
                    create: {
                        __args: {
                            input: {
                                ...convertDates(intent),
                                createdAt: intent.createdAt?.toISOString() ?? new Date().toISOString(),
                            },
                        },
                        id: true,
                        createdAt: true,
                    },
                },
            },
        };
        const confirmation = await orderApi((0, json_to_graphql_query_1.jsonToGraphQLQuery)(mutation));
        return {
            id: confirmation.orders.create.id,
            createdAt: confirmation.orders.create.createdAt,
        };
    };
}
exports.createOrderPusher = createOrderPusher;
function createOrderPaymentUpdater(apiClient) {
    return async function updaptePaymentOrder(orderId, intentOrder) {
        const intent = order_1.updateOrderInputRequest.parse(intentOrder);
        const pimApi = apiClient.pimApi;
        const mutation = {
            mutation: {
                order: {
                    update: {
                        __args: {
                            id: orderId,
                            input: convertDates(intent),
                        },
                        id: true,
                        updatedAt: true,
                    },
                },
            },
        };
        const confirmation = await pimApi((0, json_to_graphql_query_1.jsonToGraphQLQuery)(mutation));
        return {
            id: confirmation.order.update.id,
            updatedAt: confirmation.order.update.updatedAt,
        };
    };
}
exports.createOrderPaymentUpdater = createOrderPaymentUpdater;
